### PR TITLE
Special agent for fuse plugin

### PR DESCRIPTION
**Pull request by ricardomacedo**
Added:

- wato plugin
- parser and tests
- special agent and tests

### CODE REVIEW

File: checks/agent_fuse_parser [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+def agent_fuse_arguments(params: dict, hostname: str, ipaddress) -> list:
```
Comment (lisa-pi) [inline checks/agent_fuse_parser:4]: It's OK not to type this.. We will rework the argument parser at some point.

File: cmk/gui/plugins/wato/datasource_programs.py [hunk 2]
```diff
@@ -3772,3 +3772,29 @@ def _valuespec_special_agents_rabbitmq():
...
+
+def _valuespec_special_agents_fuse():
+    return Dictionary(
+        title=_("Fuse Management Central"),
+        help=
+        _("This rule set selects the special agent for Fuse Management Central "
+          "instead of the normal Check_MK agent and allows monitoring via REST API."),
+        elements=[
+            ("username", TextInput(title=_("Username"), allow_empty=False)),
+            ("password", Password(title=_("Password"), allow_empty=False)),
```
Comment (lisa-pi) [inline cmk/gui/plugins/wato/datasource_programs.py:3786]: IndividualOrStoredPassword

File: cmk/special_agents/agent_fuse.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+                        metavar="URL",
+                        help="")
+    parser.add_argument("host",
+                        metavar="HOST",
+                        help="")
+    return parser.parse_args(argv)
+
+
+def main(args: Optional[List[str]] = None) -> int:
+    if args is None:
```
Comment (lisa-pi) [inline cmk/special_agents/agent_fuse.py:196]: Add line below: 
cmk.utils.password_store.replace_passwords()

File: cmk/special_agents/agent_fuse.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    opt: argparse.Namespace = parse_arguments(args)
+
+    layout_connection: LayoutConnection = LayoutConnection(opt.url)
+    layout_request: FuseRequest  = FuseRequest(layout_connection.base_url, opt.user, opt.password)
+    layout_response: LayoutResponse = layout_request.get_layout()
+
+    summary_connection: SummaryConnection = SummaryConnection(opt.url)
+    summary_request: FuseRequest = FuseRequest(summary_connection.base_url, opt.user, opt.password)
+    summary_response: SummaryResponse = summary_request.get_summary()
+
```
Comment (lisa-pi) [inline cmk/special_agents/agent_fuse.py:207]: Move state to helper function, helps keep main() as small as possible

File: cmk/special_agents/agent_fuse.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        state = "unauth"
+    elif layout_response.code != 200 or summary_response.code != 200:
+        state = "down"
+    elif not layout_response.data:
+        state = "empty"
+
+    sys.stdout.write("<<<fuse_instance:sep(0)>>>\n")
+    sys.stdout.write("{}\n".format(state))
+
+    if state == "up":
```
Comment (lisa-pi) [inline cmk/special_agents/agent_fuse.py:219]: if state != "up":
    return

File: cmk/special_agents/agent_fuse.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+                        env_alert.warnings = int(alert["warnings"])
+                    if "link" in alert:
+                        env_alert.link = alert["link"]
+                env_alerts.append(env_alert)
+    return env_alerts
+
+
+def get_admin_alerts(layout: dict, env_admin_map: Dict[str, dict]) -> List[Alert]:
+    admin_alerts: List[Alert] = []
+    if "admin" in layout:
```
Comment (lisa-pi) [inline cmk/special_agents/agent_fuse.py:162]: if "admin" not in layout:
return

File: cmk/special_agents/agent_fuse.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+                        system_alert.warnings = int(alert["warnings"])
+                    if "link" in alert:
+                        system_alert.link = alert["link"]
+                system_alerts.append(system_alert)
+    return system_alerts
+
+
+def get_environment_alerts(layout: dict, env_alert_map: Dict[Tuple[str, str], dict]) -> List[Alert]:
+    env_alerts: List[Alert] = []
+    if "environments" in layout:
```
Comment (lisa-pi) [inline cmk/special_agents/agent_fuse.py:142]: if "environments" not in layout:
return ..

File: cmk/special_agents/agent_fuse.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        elif "envId" in alert:
+            env_alerts_map[(alert["envId"], alert["componentType"])] = alert
+        else:
+            fuse_alerts_map[alert["componentType"]] = alert
+    return SummaryStructure(system_alerts_map, env_alerts_map, fuse_alerts_map)
+
+
+def get_systems_alerts(layout: dict, system_alerts_map: Dict[Tuple[str, str], dict]) -> List[Alert]:
+    system_alerts: List[Alert] = []
+    if "systems" in layout:
```
Comment (lisa-pi) [inline cmk/special_agents/agent_fuse.py:121]: if "systems" not in layout:
return ..

**Review by lisa-pi — CHANGES_REQUESTED** [review]
Hi Ricardo, 

Thanks for splitting the commits up into smaller commits. It helps us with the reviews. 
Your special agent is well-written. In addition to my in-line comments, I have a couple of points

- Please name your special agent "fuse_management_central" (I have this from the title of the rulespec). More accurate names help keep the agent names unique and help us understand what they're about. 
- The naming convention for the argument parser is agent_\<agent name\>
- You don't need to type everything. What we'd like typed are the function signatures (the parameters that you feed in to your functions and their return value). Typing within the function code is only necessary when you define variables for which mypy can't derive a type. 
- For complex data structures (e.g. Dict[Tuple[str, str],....]) please try to define more accurate types. For this, most of the time classes or NamedTuples can be useful.
- What is the benefit of defining the SummaryStructure? Why not leave the objects "system_alerts", "env_alerts" and "fuse_alerts" as separate objects with their own properties? 
- In your tests, could you please use more accurate names for your input parameters? They are currently always "params" and "result". More meaningful names would help me understand better what you're testing.

You are currently missing the following:
- A stub in the directory agents/special/
- An entry in the file tests/unit/agents/special/test_special_agent_args.py, analogous to the other agents

If you have any questions, please feel free to get in touch. Happy to do a video call if that helps. 

Cheers,
Lisa

**Comment by ricardomacedo** [comment]
Hi Lisa,

I just commited the changes.

Cheers,
Ricardo

**Comment by ricardomacedo** [comment]
Hi @lisa-pi,

Can you give us some help to help to fix the failing check?

Cheers

File: tests/unit/checks/test_agent_fuse_management_central.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+import pytest
+
+from testlib import SpecialAgent
```
Comment (svenpanne) [inline tests/unit/checks/test_agent_fuse_management_central.py:6]: Importing from `tests.testlib` should work.

**Comment by ricardomacedo** [comment]
Hi @lisa-pi and @svenpanne,

One check failed, can you help us find out the problem?

Cheers,
Ricardo Macedo

**Comment by lisa-pi** [comment]
> Hi @lisa-pi and @svenpanne,
> 
> One check failed, can you help us find out the problem?
> 
> Cheers,
> Ricardo Macedo

Hi @ricardomacedo, you must add a test case for your special agent to the test file
`unit/agents/special/test_special_agent_args.py`

**Comment by ricardomacedo** [comment]
> > Hi @lisa-pi and @svenpanne,
> > One check failed, can you help us find out the problem?
> > Cheers,
> > Ricardo Macedo
> 
> Hi @ricardomacedo, you must add a test case for your special agent to the test file
> `unit/agents/special/test_special_agent_args.py`

Hi @lisa-pi,

We have the following files:

- `checks/agent_fuse_management_central` with the tests in `tests/unit/checks/test_agent_fuse_management_central.py`
- `cmk/special_agents/agent_fuse_management_central.py` with the tests in `tests/unit/cmk/special_agents/test_agent_fuse_management_central.py`
- `agents/special/agent_fuse_management_central` is just a stub

And then we also changed the file `cmk/gui/plugins/wato/datasource_programs.py`.

In the file `tests/unit/checks/test_agent_fuse_management_central.py` we have the test case for the special agent, is there something wrong or missing?

**Comment by lisa-pi** [comment]
> unit/agents/special/test_special_agent_args.py

Hi @ricardomacedo, the file `unit/agents/special/test_special_agent_args.py` is an existing test file. You also need to add a test case there for your special agent. Have a look at the file, there are many examples.

**Comment by ricardomacedo** [comment]
Hi @lisa-pi,

Can you help us fixing the test case for the special agent in the test file `unit/agents/special/test_special_agent_args.py`?

Cheers

**Comment by lisa-pi** [comment]
> Hi @lisa-pi,
> 
> Can you help us fixing the test case for the special agent in the test file `unit/agents/special/test_special_agent_args.py`?
> 
> Cheers

Hi @ricardomacedo, your special agent should have an additional argument: "--debug". See "Details": 
```
         msg = "Special agents should support the argument '--debug'"
        # This also ensures that the parse_arguments function indeed expects
        # sys.argv[1:], i.e. the first element omitted.
>       assert parse_arguments(['--debug'] + required_args).debug is True, msg
```

**Comment by ricardomacedo** [comment]
> > Hi @lisa-pi,
> > Can you help us fixing the test case for the special agent in the test file `unit/agents/special/test_special_agent_args.py`?
> > Cheers
> 
> Hi @ricardomacedo, your special agent should have an additional argument: "--debug". See "Details":
> 
> ```
>          msg = "Special agents should support the argument '--debug'"
>         # This also ensures that the parse_arguments function indeed expects
>         # sys.argv[1:], i.e. the first element omitted.
> >       assert parse_arguments(['--debug'] + required_args).debug is True, msg
> ```

Hi @lisa-pi, can you point us one example of a special agent accepting the argument --debug?

**Comment by lisa-pi** [comment]
> > > Hi @lisa-pi,
> > > Can you help us fixing the test case for the special agent in the test file `unit/agents/special/test_special_agent_args.py`?
> > > Cheers
> > 
> > 
> > Hi @ricardomacedo, your special agent should have an additional argument: "--debug". See "Details":
> > ```
> >          msg = "Special agents should support the argument '--debug'"
> >         # This also ensures that the parse_arguments function indeed expects
> >         # sys.argv[1:], i.e. the first element omitted.
> > >       assert parse_arguments(['--debug'] + required_args).debug is True, msg
> > ```
> 
> Hi @lisa-pi, can you point us one example of a special agent accepting the argument --debug?

Hi @ricardomacedo , any special agent should have it. E.g. `cmk/special_agents/agent_siemens_plc.py`

**Comment by lisa-pi** [comment]
Hi @ricardomacedo , it appears that something got mixed up in your last commit. Please delete it and rebase your branch on our master branch instead (make sure to do a pull from master before you rebase).
Note: we are now using black as our formatting tool. This is why you are experiencing merge conflicts. Please use black to format your code going forward. 
Sorry for the inconvenience!

**Comment by ricardomacedo** [comment]
Hi @lisa-pi,

Branch rebased, files formated and all checks are passing.

Cheers

File: cmk/special_agents/agent_fuse_management_central.py [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+    parser.add_argument("password", metavar="PASSWORD", help="")
+    parser.add_argument("url", metavar="URL", help="")
+    parser.add_argument("host", metavar="HOST", help="")
+    parser.add_argument(
+        "--debug", action="store_true", help="Debug mode: let Python exceptions raise through"
+    )
+    return parser.parse_args(argv)
+
+
+def get_state(layout_response_code: int, summary_response_code: int, data) -> str:
```
Comment (lisa-pi) [inline cmk/special_agents/agent_fuse_management_central.py:211]: I think you want to make this a service later. In that case, it is better to give out the raw information here and let the check decide what to do with it later. It means more flexibility and also that the evaluation is always in the same place.

File: cmk/special_agents/agent_fuse_management_central.py [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+    layout_request: FuseRequest = FuseRequest(layout_connection.base_url, opt.user, opt.password)
+    layout_response: LayoutResponse = layout_request.get_layout()
+
+    summary_connection: SummaryConnection = SummaryConnection(opt.url)
+    summary_request: FuseRequest = FuseRequest(summary_connection.base_url, opt.user, opt.password)
+    summary_response: SummaryResponse = summary_request.get_summary()
+
+    state: str = get_state(layout_response.code, summary_response.code, layout_response.data)
+
+    sys.stdout.write("<<<fuse_instance:sep(0)>>>\n")
```
Comment (lisa-pi) [inline cmk/special_agents/agent_fuse_management_central.py:238]: `fuse_management_central_instance`

File: cmk/special_agents/agent_fuse_management_central.py [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+    if state != "up":
+        return 0
+
+    layout: dict = layout_response.data
+    summary: list = summary_response.data
+
+    summary_structure: SummaryStructure = get_summary_structure(summary)
+
+    system_alerts: List[Alert] = get_systems_alerts(layout, summary_structure.system_alerts)
+    sys.stdout.write("<<<fuse_system_alerts:sep(0)>>>\n")
```
Comment (lisa-pi) [inline cmk/special_agents/agent_fuse_management_central.py:250]: `fuse_management_central_system_alerts`

File: cmk/special_agents/agent_fuse_management_central.py [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+    sys.stdout.write("{}\n".format(json.dumps([alert.__dict__ for alert in system_alerts])))
+
+    env_alerts: List[Alert] = get_environment_alerts(layout, summary_structure.env_alerts)
+    sys.stdout.write("<<<fuse_env_alerts:sep(0)>>>\n")
```
Comment (lisa-pi) [inline cmk/special_agents/agent_fuse_management_central.py:254]: `fuse_management_central_env_alerts`

File: cmk/special_agents/agent_fuse_management_central.py [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+    sys.stdout.write("{}\n".format(json.dumps([alert.__dict__ for alert in env_alerts])))
+
+    admin_alerts: List[Alert] = get_admin_alerts(layout, summary_structure.fuse_alerts)
+    sys.stdout.write("<<<fuse_admin_alerts:sep(0)>>>\n")
```
Comment (lisa-pi) [inline cmk/special_agents/agent_fuse_management_central.py:258]: `fuse_management_central_admin_alerts`

File: cmk/special_agents/agent_fuse_management_central.py [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+    def __init__(self, url: str) -> None:
+        self.base_url: str = "%s/layout" % url
+
+
+class SummaryConnection:
+    def __init__(self, url: str) -> None:
+        self.base_url: str = "%s/summary" % url
+
+
+class FuseRequest:
```
Comment (lisa-pi) [inline cmk/special_agents/agent_fuse_management_central.py:92]: I think we don't need this class. See below.

File: cmk/special_agents/agent_fuse_management_central.py [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+
+def main(args: Optional[List[str]] = None) -> int:
+    if args is None:
+        cmk.utils.password_store.replace_passwords()
+        args = sys.argv[1:]
+    opt: argparse.Namespace = parse_arguments(args)
+
+    layout_connection: LayoutConnection = LayoutConnection(opt.url)
+    layout_request: FuseRequest = FuseRequest(layout_connection.base_url, opt.user, opt.password)
+    layout_response: LayoutResponse = layout_request.get_layout()
```
Comment (lisa-pi) [inline cmk/special_agents/agent_fuse_management_central.py:230]: `layout_response = requests.get(f'{opt.url}/layout', auth=(opt.user, opt.password))`

For error handling: when a special agent returns a non-zero exit code, the first line of the error message written to stderr is shown by the "Check_MK" service and it returns a state CRIT. You can use this fact for your error handling here. Eg.

```
try: 
    layout_response = requests.get(f'{opt.url}/layout', auth=(opt.user, opt.password))
except requests.exceptions.RequestException:
    print('Some useful error text', file=sys.stderr)
    raise SystemExit(1)
```

File: cmk/special_agents/agent_fuse_management_central.py [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+
+    summary_connection: SummaryConnection = SummaryConnection(opt.url)
+    summary_request: FuseRequest = FuseRequest(summary_connection.base_url, opt.user, opt.password)
+    summary_response: SummaryResponse = summary_request.get_summary()
```
Comment (lisa-pi) [inline cmk/special_agents/agent_fuse_management_central.py:234]: See above

**Review by lisa-pi — CHANGES_REQUESTED** [review]
Hi @ricardomacedo , 
Thanks for your PR. Here's a few comments. If you've any questions, let me know. 
Cheers, 
Lisa

**Comment by jherbel** [comment]
Closed this due to inactivity

**Pull request closed without merging**